awx/main/models/unified_jobs.py

```diff
@@ -842,15 +843,16 @@ def save(self, *args, **kwargs):
             if 'finished' not in update_fields:
                 update_fields.append('finished')
 
+        dq = decimal.Decimal('1.000')
+        if self.elapsed is None:
+            self.elapsed = decimal.Decimal(0.0).quantize(dq)
+
         # If we have a start and finished time, and haven't already calculated
         # out the time that elapsed, do so.
-        if self.started and self.finished and not self.elapsed:
+        if self.started and self.finished and self.elapsed == 0.0:
             td = self.finished - self.started
-            elapsed = (td.microseconds + (td.seconds + td.days * 24 * 3600) * 10 ** 6) / (10 ** 6 * 1.0)
-        else:
-            elapsed = 0.0
-        if self.elapsed != elapsed:
-            self.elapsed = str(elapsed)
+            elapsed = decimal.Decimal(td.total_seconds())
+            self.elapsed = elapsed.quantize(dq)
```

Comment:
I would only have 1 minor tweak to suggest - before hitting this line, add another conditional line like

```python
if elapsed != self.elapsed:
```

Otherwise, it looks like every save after a job finished would add this `elapsed` field to the payload for updating the row. It would be preferable and intuitive that it updates the field exactly once, on the first save after the `finished` field is filled in.

Reply:
I think `and self.elapsed == 0.0` in the conditional should cover this. Once we hit this code block, self.elapsed will be non-zero, and any subsequent calls to `.save()` should skip this code.

To confirm this, we can add log messages for when we save and when we update elapsed. You can see it saves three times, but updates elapsed once.

```bash
tools_awx_1 | awx-dispatcher stdout | 2021-04-09 22:01:51,814 DEBUG    [4bcbe8ea] awx.main.models.unified_jobs ====== in save
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-04-09 22:01:51,815 DEBUG    [4bcbe8ea] awx.main.models.unified_jobs ===== updating elapsed
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-04-09 22:01:51,901 DEBUG    [4bcbe8ea] awx.analytics.job_lifecycle job-752 finalize run
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-04-09 22:01:51,904 DEBUG    [4bcbe8ea] awx.main.models.unified_jobs ====== in save
tools_awx_1 | awx-dispatcher stdout | 
tools_awx_1 | awx-dispatcher stdout | 2021-04-09 22:01:51,923 DEBUG    [4bcbe8ea] awx.main.models.unified_jobs ====== in save
```